### ISSUE TITLE
训练时梯度累积有问题？

### ISSUE THREAD

**Issue by Vitalihh** [modelscope/DiffSynth-Studio#1388]
runner.py中的launch_training_task()中，先清空了梯度再更新参数，这不会导致梯度累计失去作用吗？
    model, optimizer, dataloader, scheduler = accelerator.prepare(model, optimizer, dataloader, scheduler)
    initialize_deepspeed_gradient_checkpointing(accelerator)
    for epoch_id in range(num_epochs):
        for data in tqdm(dataloader):
            with accelerator.accumulate(model):
                optimizer.zero_grad()
                if dataset.load_from_cache:
                    loss = model({}, inputs=data)
                else:
                    loss = model(data)
                accelerator.backward(loss)
                optimizer.step()
                model_logger.on_step_end(accelerator, model, save_steps, loss=loss)
                scheduler.step()
        if save_steps is None:
            model_logger.on_epoch_end(accelerator, model, epoch_id)
    model_logger.on_training_end(accelerator, model, save_steps)
是否应该改为：
 for epoch_id in range(num_epochs):
        for data in tqdm(dataloader):
            with accelerator.accumulate(model):
                loss = model(data)
                accelerator.backward(loss)
                if accelerator.sync_gradients and max_grad_norm > 0:
                    accelerator.clip_grad_norm_(model.parameters(), max_grad_norm)
                optimizer.step()
                model_logger.on_step_end(accelerator, model, save_steps, loss=loss)
                scheduler.step()
                optimizer.zero_grad()
        if save_steps is None:
            model_logger.on_epoch_end(accelerator, model, epoch_id)
    model_logger.on_training_end(accelerator, model, save_steps)
先更新参数后清空梯度

**Comment by mi804** [comment]
这是with accelerator.accumulate(model):下的标准模型训练流程，梯度更新是正常的

**Comment by firefighter-eric** [comment]
我看了一下这里的训练代码，感觉 `optimizer.zero_grad()` 放的位置不太对，可能会导致 `gradient_accumulation` 实际上没有生效。

现在的代码大概是这样：

```python

for data in tqdm(dataloader):

    with accelerator.accumulate(model):

        optimizer.zero_grad()

        if dataset.load_from_cache:

            loss = model({}, inputs=data)

        else:

            loss = model(data)

        accelerator.backward(loss)

        optimizer.step()

        model_logger.on_step_end(accelerator, model, save_steps, loss=loss)

        scheduler.step()

```

但按照 Accelerate 常见的写法，`zero_grad()` 一般应该放在 `optimizer.step()` 后面，而不是一开始就清掉：

```python

for data in dataloader:

    with accelerator.accumulate(model):

        loss = ...

        accelerator.backward(loss)

        optimizer.step()

        scheduler.step()

        optimizer.zero_grad()

```

我担心现在这段代码在 `gradient_accumulation_steps > 1` 的时候会有问题。

因为在做梯度累积时，前几个 micro-step 的梯度本来应该一直留着，等累计够了再一起更新。  

但如果在 `accumulate()` 一进来就先 `zero_grad()`，那最后一个 micro-step 在真正 `step()` 之前，前面累积的梯度可能已经被清掉了。这样一来，最终更新可能只用了最后一个 micro-batch 的梯度，而不是累计后的梯度。

表面上看 loss 还是可能会下降，所以这个问题不一定很容易第一眼发现，但 accumulation 的效果可能已经不对了。

我理解这里更稳妥的写法应该是：

```python

for data in tqdm(dataloader):

    with accelerator.accumulate(model):

        if dataset.load_from_cache:

            loss = model({}, inputs=data)

        else:

            loss = model(data)

        accelerator.backward(loss)

        optimizer.step()

        scheduler.step()

        optimizer.zero_grad()

```

**Comment by firefighter-eric** [comment]
@Artiprocher @mi804 可以帮忙看下吗
我提了一个pr，在 https://github.com/modelscope/DiffSynth-Studio/pull/1404

**Comment by Vitalihh** [comment]
@firefighter-eric 我提出的问题，你提PR？
